edb/edgeql/compiler/casts.py

```diff
@@ -356,6 +364,46 @@ def _find_cast(
         return None
 
 
+def _cast_json_to_tuple(
+        ir_set: irast.Set,
+        orig_stype: s_types.Type,
+        new_stype: s_types.Tuple, *,
+        srcctx: Optional[parsing.ParserContext],
+        ctx: context.ContextLevel) -> irast.Set:
+
+    new_subtypes = dict(new_stype.iter_subtypes(ctx.env.schema))
+
+    with ctx.new() as subctx:
+        subctx.anchors = subctx.anchors.copy()
+        source_alias = subctx.aliases.get('a')
+        subctx.anchors[source_alias] = ir_set
+
+        elements = []
+        for new_el_name, new_st in new_subtypes.items():
```

Comment:
Why not `for ... in new_stype.iter_subtypes(ctx.env.schema)`?  That would avoid allocating a dictionary just for iteration.

Reply:
Oh, yeah, obviously. (I had copy-pasted and adapted from some other tuple casting code that made a dict so it could ask the length, it looks like)